File: nestjs/src/booking/core/domain/entities/room.entity.ts
import { Entity } from "@/shared/domain/entities/entity";
import { Price } from "../value-objects/price.valueobject";

export type RoomProps = {
    id: number;
    name: string;
    level: number;
    inMaintenance: boolean;
    price: Price;
}

export default class RoomEntity extends Entity<RoomProps> {
    constructor(public readonly props: RoomProps) {
        super(props);
    }

    isAvailable(): boolean {
        return !this.props.inMaintenance || this.hasGuest() ? false : true;
    }

    hasGuest(): boolean {
        return true;
    }
}

File: go/internal/booking-service/core/domain/entities/room.entity.go
package entities

import valueobjects "book/internal/booking-service/core/domain/value-objects"

type Room struct {
	Id            uint `gorm:"primaryKey"`
	Name          string
	Level         int
	InMaintenance bool
	Price         valueobjects.Price `gorm:"embedded"`
}

func (r Room) IsAvailable() bool {
	if !r.InMaintenance || r.HasGuest() {
		return false
	}
	return true
}

func (r Room) HasGuest() bool {
	return true
}
